File: app/src/test/java/bj/discogsbrowser/release/ReleasePresenterTest.java
package bj.discogsbrowser.release;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import bj.discogsbrowser.greendao.DaoManager;
import bj.discogsbrowser.model.release.Release;
import bj.discogsbrowser.network.DiscogsInteractor;
import bj.discogsbrowser.utils.ArtistsBeautifier;
import bj.discogsbrowser.utils.schedulerprovider.TestSchedulerProvider;
import io.reactivex.Single;
import io.reactivex.schedulers.TestScheduler;

import static junit.framework.Assert.assertEquals;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

/**
 * Created by Josh Laird on 10/05/2017.
 */
@RunWith(MockitoJUnitRunner.class)
public class ReleasePresenterTest
{
    private ReleasePresenter presenter;
    @Mock ReleaseController controller;
    @Mock DiscogsInteractor discogsInteractor;
    private TestScheduler testScheduler = new TestScheduler();
    @Mock DaoManager daoManager;
    @Mock ArtistsBeautifier artistsBeautifier;
    private ReleaseFactory releaseFactory = new ReleaseFactory();
    private String id = "123";

    @Before
    public void setUp()
    {
        presenter = new ReleasePresenter(controller, discogsInteractor,
                new TestSchedulerProvider(testScheduler), daoManager, artistsBeautifier);
    }

    @After
    public void tearDown()
    {
        verifyNoMoreInteractions(controller, discogsInteractor,
                daoManager, artistsBeautifier);
    }

//    @Test
//    public void setupRecyclerView_setsUpRecyclerView()
//    {
//        Context mockCtx = mock(Context.class);
//        RecyclerView mockRv = mock(RecyclerView.class);
//        EpoxyControllerAdapter mockAdapter = mock(EpoxyControllerAdapter.class);
//        String title = "yedawg";
//        when(controller.getAdapter()).thenReturn(mockAdapter);
//
//        presenter.setupRecyclerView(mockCtx, mockRv, title);
//
//        verify(mockRv).setLayoutManager(any(LinearLayoutManager.class));
//        verify(mockRv).setAdapter(mockAdapter);
//        verify(controller).getAdapter();
//        verify(controller).setTitle(title);
//        verify(controller).requestModelBuild();
//    }

    @Test
    public void getReleaseAndLabelDetailsError_displaysError()
    {
        when(discogsInteractor.fetchReleaseDetails(id)).thenReturn(Single.error(new Throwable()));

        presenter.getReleaseAndLabelDetails(id);
        testScheduler.triggerActions();

        verify(discogsInteractor).fetchReleaseDetails(id);
        verify(controller).setReleaseLoading(true);
        verify(controller).setReleaseError(true);
    }

    @Test
    public void fetchReleaseListingsError_displaysError() throws IOException
    {
        when(discogsInteractor.getReleaseMarketListings(id)).thenReturn(Single.error(new Throwable()));

        presenter.fetchReleaseListings(id);
        testScheduler.triggerActions();

        verify(discogsInteractor).getReleaseMarketListings(id);
        verify(controller).setListingsLoading(true);
        verify(controller).setReleaseListingsError();
    }

    @Test
    public void checkCollectionError_displaysError() throws IOException
    {
        Release releaseNoLabelNoneForSale = releaseFactory.getReleaseNoLabelNoneForSaleNoTracklistNoVideos();
        when(controller.getRelease()).thenReturn(releaseNoLabelNoneForSale);
        when(discogsInteractor.checkIfInCollection(controller, releaseNoLabelNoneForSale)).thenReturn(Single.error(new Throwable()));

        presenter.checkCollectionWantlist();
        testScheduler.triggerActions();

        verify(controller, times(1)).getRelease();
        verify(discogsInteractor).checkIfInCollection(controller, releaseNoLabelNoneForSale);
        verify(controller).setCollectionWantlistError(true);
    }

    @Test
    public void checkCollectionValid_displaysValid() throws IOException
    {
        Release releaseNoLabelNoneForSale = releaseFactory.getReleaseNoLabelNoneForSaleNoTracklistNoVideos();
        when(controller.getRelease()).thenReturn(releaseNoLabelNoneForSale);
        when(discogsInteractor.checkIfInCollection(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just(Collections.emptyList()));
        when(discogsInteractor.checkIfInWantlist(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just(Collections.emptyList()));

        presenter.checkCollectionWantlist();
        testScheduler.triggerActions();

        verify(controller, times(2)).getRelease();
        verify(discogsInteractor).checkIfInCollection(controller, releaseNoLabelNoneForSale);
        verify(discogsInteractor).checkIfInWantlist(controller, releaseNoLabelNoneForSale);
        verify(controller).setCollectionWantlistChecked(true);
    }

    @Test
    public void getReleaseNoLabelNoneForSale_displaysRelease()
    {
        Release releaseNoLabelNoneForSale = releaseFactory.getReleaseNoLabelNoneForSaleNoTracklistNoVideos();
        Single<Release> releaseSingle = Single.just(releaseNoLabelNoneForSale);
        when(discogsInteractor.fetchReleaseDetails(id)).thenReturn(releaseSingle);
        ArgumentCaptor<ArrayList> arrayListArgumentCaptor = ArgumentCaptor.forClass(ArrayList.class);
        when(discogsInteractor.checkIfInCollection(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just(new ArrayList<>()));
        when(discogsInteractor.checkIfInWantlist(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just(new ArrayList<>()));
        when(controller.getRelease()).thenReturn(releaseNoLabelNoneForSale);

        presenter.getReleaseAndLabelDetails(id);
        testScheduler.triggerActions();

        verify(discogsInteractor).fetchReleaseDetails(id);
        verify(controller).setReleaseLoading(true);
        verify(daoManager).storeViewedRelease(releaseNoLabelNoneForSale, artistsBeautifier);
        verify(controller).setRelease(releaseNoLabelNoneForSale);
        verify(controller).setReleaseListings(arrayListArgumentCaptor.capture());
        assertEquals(arrayListArgumentCaptor.getValue().size(), 0);
        verify(discogsInteractor, times(1)).checkIfInCollection(controller, releaseNoLabelNoneForSale);
        verify(discogsInteractor, times(1)).checkIfInWantlist(controller, releaseNoLabelNoneForSale);
        verify(controller, times(2)).getRelease();
        verify(controller).setCollectionWantlistChecked(true);
    }

    @Test
    public void getReleaseWithLabelNoneForSale_displaysRelease()
    {
        Release releaseWithLabelNoneForSale = releaseFactory.getReleaseWithLabelNoneForSale();
        Single<Release> releaseSingle = Single.just(releaseWithLabelNoneForSale);
        when(discogsInteractor.fetchReleaseDetails(id)).thenReturn(releaseSingle);
        when(discogsInteractor.fetchLabelDetails(id)).thenReturn(Single.just(releaseFactory.getLabelDetails()));
        ArgumentCaptor<ArrayList> arrayListArgumentCaptor = ArgumentCaptor.forClass(ArrayList.class);
        when(discogsInteractor.checkIfInCollection(controller, releaseWithLabelNoneForSale)).thenReturn(Single.just(new ArrayList<>()));
        when(discogsInteractor.checkIfInWantlist(controller, releaseWithLabelNoneForSale)).thenReturn(Single.just(new ArrayList<>()));
        when(controller.getRelease()).thenReturn(releaseWithLabelNoneForSale);

        presenter.getReleaseAndLabelDetails(id);
        testScheduler.triggerActions();

        verify(discogsInteractor).fetchReleaseDetails(id);
        verify(controller).setReleaseLoading(true);
        verify(discogsInteractor).fetchLabelDetails(id);
        verify(daoManager).storeViewedRelease(releaseWithLabelNoneForSale, artistsBeautifier);
        verify(controller).setRelease(releaseWithLabelNoneForSale);
        verify(controller).setReleaseListings(arrayListArgumentCaptor.capture());
        assertEquals(arrayListArgumentCaptor.getValue().size(), 0);
        verify(discogsInteractor, times(1)).checkIfInCollection(controller, releaseWithLabelNoneForSale);
        verify(discogsInteractor, times(1)).checkIfInWantlist(controller, releaseWithLabelNoneForSale);
        verify(controller, times(2)).getRelease();
        verify(controller).setCollectionWantlistChecked(true);
    }

    @Test
    public void getReleaseWithLabelAndListings_displaysReleaseAndListings() throws IOException
    {
        Release releaseWithLabelSomeForSale = releaseFactory.getReleaseWithLabelSomeForSale();
        Single<Release> releaseSingle = Single.just(releaseWithLabelSomeForSale);
        when(discogsInteractor.fetchReleaseDetails(id)).thenReturn(releaseSingle);
        when(discogsInteractor.fetchLabelDetails(id)).thenReturn(Single.just(releaseFactory.getLabelDetails()));
        ArgumentCaptor<List> listArgumentCaptor = ArgumentCaptor.forClass(ArrayList.class);
        when(discogsInteractor.getReleaseMarketListings(id)).thenReturn(Single.just(releaseFactory.getFourScrapeListings()));

        when(discogsInteractor.checkIfInCollection(controller, releaseWithLabelSomeForSale)).thenReturn(Single.just(new ArrayList<>()));
        when(discogsInteractor.checkIfInWantlist(controller, releaseWithLabelSomeForSale)).thenReturn(Single.just(new ArrayList<>()));
        when(controller.getRelease()).thenReturn(releaseWithLabelSomeForSale);

        presenter.getReleaseAndLabelDetails(id);
        testScheduler.triggerActions();

        verify(discogsInteractor).fetchReleaseDetails(id);
        verify(controller).setReleaseLoading(true);
        verify(discogsInteractor).fetchLabelDetails(id);
        verify(daoManager).storeViewedRelease(releaseWithLabelSomeForSale, artistsBeautifier);
        verify(controller).setRelease(releaseWithLabelSomeForSale);
        verify(controller).setReleaseListings(listArgumentCaptor.capture());
        assertEquals(listArgumentCaptor.getValue().size(), 4);
        verify(discogsInteractor, times(1)).checkIfInCollection(controller, releaseWithLabelSomeForSale);
        verify(discogsInteractor, times(1)).checkIfInWantlist(controller, releaseWithLabelSomeForSale);
        verify(discogsInteractor).getReleaseMarketListings(id);
        verify(controller, times(2)).getRelease();
        verify(controller).setCollectionWantlistChecked(true);
        verify(controller).setListingsLoading(true);
    }
}

File: app/src/test/java/bj/vinylbrowser/release/ReleasePresenterTest.kt
package bj.vinylbrowser.release

import bj.vinylbrowser.greendao.DaoManager
import bj.vinylbrowser.model.collection.CollectionRelease
import bj.vinylbrowser.model.common.Label
import bj.vinylbrowser.model.listing.ScrapeListing
import bj.vinylbrowser.model.release.Release
import bj.vinylbrowser.model.wantlist.Want
import bj.vinylbrowser.network.DiscogsInteractor
import bj.vinylbrowser.utils.ArtistsBeautifier
import bj.vinylbrowser.utils.schedulerprovider.TestSchedulerProvider
import com.nhaarman.mockito_kotlin.*
import io.reactivex.Single
import io.reactivex.schedulers.TestScheduler
import junit.framework.Assert.assertEquals
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentCaptor
import org.mockito.MockitoAnnotations.initMocks
import org.mockito.junit.MockitoJUnitRunner
import java.io.IOException

/**
 * Created by Josh Laird on 22/05/2017.
 */
@RunWith(MockitoJUnitRunner::class)
class ReleasePresenterTest {
    lateinit var presenter: ReleasePresenter
    val controller: ReleaseEpxController = mock()
    val discogsInteractor: DiscogsInteractor = mock()
    val testScheduler = TestScheduler()
    val daoManager: DaoManager = mock()
    val artistsBeautifier: ArtistsBeautifier = mock()
    val id = "123"

    @Before
    fun setUp() {
        initMocks(this)
        presenter = ReleasePresenter(controller, discogsInteractor,
                TestSchedulerProvider(testScheduler), daoManager, artistsBeautifier)
    }

    @After
    fun tearDown() {
        verifyNoMoreInteractions(controller, discogsInteractor,
                daoManager, artistsBeautifier)
    }

    @Test
    fun getReleaseAndLabelDetailsError_displaysError() {
        whenever(discogsInteractor.fetchReleaseDetails(id)).thenReturn(Single.error<Release>(Throwable()))

        presenter.fetchArtistDetails(id)
        testScheduler.triggerActions()

        verify(discogsInteractor).fetchReleaseDetails(id)
        verify(controller).setReleaseLoading(true)
        verify(controller).setReleaseError(true)
    }

    @Test
    @Throws(IOException::class)
    fun fetchReleaseListingsError_displaysError() {
        whenever(discogsInteractor.getReleaseMarketListings(id)).thenReturn(Single.error<List<ScrapeListing>>(Throwable()))

        presenter.fetchReleaseListings(id)
        testScheduler.triggerActions()

        verify(discogsInteractor).getReleaseMarketListings(id)
        verify(controller).setListingsLoading(true)
        verify(controller).setReleaseListingsError()
    }

    @Test
    @Throws(IOException::class)
    fun checkCollectionError_displaysError() {
        val releaseNoLabelNoneForSale = ReleaseFactory.getReleaseNoLabelNoneForSaleNoTracklistNoVideos(id)
        whenever(controller.release).thenReturn(releaseNoLabelNoneForSale)
        whenever(discogsInteractor.checkIfInCollection(controller, releaseNoLabelNoneForSale)).thenReturn(Single.error<List<CollectionRelease>>(Throwable()))

        presenter.checkCollectionWantlist()
        testScheduler.triggerActions()

        verify(controller, times(1)).getRelease()
        verify(discogsInteractor).checkIfInCollection(controller, releaseNoLabelNoneForSale)
        verify(controller).setCollectionWantlistError(true)
    }

    @Test
    @Throws(IOException::class)
    fun checkCollectionValid_displaysValid() {
        val releaseNoLabelNoneForSale = ReleaseFactory.getReleaseNoLabelNoneForSaleNoTracklistNoVideos(id)
        whenever(controller.release).thenReturn(releaseNoLabelNoneForSale)
        whenever(discogsInteractor.checkIfInCollection(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just<List<CollectionRelease>>(emptyList()))
        whenever(discogsInteractor.checkIfInWantlist(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just<List<Want>>(emptyList()))

        presenter.checkCollectionWantlist()
        testScheduler.triggerActions()

        verify(controller, times(2)).release
        verify(discogsInteractor).checkIfInCollection(controller, releaseNoLabelNoneForSale)
        verify(discogsInteractor).checkIfInWantlist(controller, releaseNoLabelNoneForSale)
        verify(controller).setCollectionWantlistChecked(true)
    }

    @Test
    fun getReleaseNoLabelNoneForSale_displaysRelease() {
        val releaseNoLabelNoneForSale = ReleaseFactory.getReleaseNoLabelNoneForSaleNoTracklistNoVideos(id)
        val releaseSingle = Single.just(releaseNoLabelNoneForSale)
        whenever(discogsInteractor.fetchReleaseDetails(id)).thenReturn(releaseSingle)
        val arrayListArgumentCaptor = ArgumentCaptor.forClass(MutableList::class.java)
        whenever(discogsInteractor.checkIfInCollection(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just<List<CollectionRelease>>(emptyList()))
        whenever(discogsInteractor.checkIfInWantlist(controller, releaseNoLabelNoneForSale)).thenReturn(Single.just<List<Want>>(emptyList()))
        whenever(controller.release).thenReturn(releaseNoLabelNoneForSale)

        presenter.fetchArtistDetails(id)
        testScheduler.triggerActions()

        verify(discogsInteractor).fetchReleaseDetails(id)
        verify(controller).setReleaseLoading(true)
        verify(daoManager).storeViewedRelease(releaseNoLabelNoneForSale, artistsBeautifier)
        verify(controller).release = releaseNoLabelNoneForSale
        verify(controller).setReleaseListings(capture(arrayListArgumentCaptor) as MutableList<ScrapeListing>?)
        assertEquals(arrayListArgumentCaptor.value.size, 0)
        verify(discogsInteractor, times(1)).checkIfInCollection(controller, releaseNoLabelNoneForSale)
        verify(discogsInteractor, times(1)).checkIfInWantlist(controller, releaseNoLabelNoneForSale)
        verify(controller, times(2)).release
        verify(controller).setCollectionWantlistChecked(true)
    }

    @Test
    fun getReleaseWithLabelNoneForSale_displaysRelease() {
        val releaseWithLabelNoneForSale = ReleaseFactory.buildReleaseWithLabelNoneForSale(id)
        val releaseSingle = Single.just(releaseWithLabelNoneForSale)
        whenever(discogsInteractor.fetchReleaseDetails(id)).thenReturn(releaseSingle)
        whenever(discogsInteractor.fetchLabelDetails(any<String>())).thenReturn(Single.just(Label())) //TODO: Change from any(). Java/Kotlin interaction is causing invalid data being returned
        val arrayListArgumentCaptor = ArgumentCaptor.forClass(MutableList::class.java)
        whenever(discogsInteractor.checkIfInCollection(controller, releaseWithLabelNoneForSale)).thenReturn(Single.just<List<CollectionRelease>>(emptyList()))
        whenever(discogsInteractor.checkIfInWantlist(controller, releaseWithLabelNoneForSale)).thenReturn(Single.just<List<Want>>(emptyList()))
        whenever(controller.release).thenReturn(releaseWithLabelNoneForSale)

        presenter.fetchArtistDetails(id)
        testScheduler.triggerActions()

        verify(discogsInteractor).fetchReleaseDetails(id)
        verify(controller).setReleaseLoading(true)
        verify(discogsInteractor).fetchLabelDetails(any<String>()) //TODO: See above
        verify(daoManager).storeViewedRelease(releaseWithLabelNoneForSale, artistsBeautifier)
        verify(controller).release = releaseWithLabelNoneForSale
        verify(controller).setReleaseListings(arrayListArgumentCaptor.capture() as MutableList<ScrapeListing>?)
        assertEquals(arrayListArgumentCaptor.value.size, 0)
        verify(discogsInteractor, times(1)).checkIfInCollection(controller, releaseWithLabelNoneForSale)
        verify(discogsInteractor, times(1)).checkIfInWantlist(controller, releaseWithLabelNoneForSale)
        verify(controller, times(2)).release
        verify(controller).setCollectionWantlistChecked(true)
    }

    @Test
    @Throws(IOException::class)
    fun getReleaseWithLabelAndListings_displaysReleaseAndListings() {
        val releaseWithLabelSomeForSale = ReleaseFactory.buildReleaseWithLabelSomeForSale(id)
        val releaseSingle = Single.just(releaseWithLabelSomeForSale)
        whenever(discogsInteractor.fetchReleaseDetails(id)).thenReturn(releaseSingle)
        whenever(discogsInteractor.fetchLabelDetails(id)).thenReturn(Single.just(Label()))
        val listArgumentCaptor = ArgumentCaptor.forClass(MutableList::class.java)
        whenever(discogsInteractor.getReleaseMarketListings(id)).thenReturn(Single.just<List<ScrapeListing>>(ScrapeListFactory.buildFourEmptyScrapeListing()))

        whenever(discogsInteractor.checkIfInCollection(controller, releaseWithLabelSomeForSale)).thenReturn(Single.just<List<CollectionRelease>>(emptyList()))
        whenever(discogsInteractor.checkIfInWantlist(controller, releaseWithLabelSomeForSale)).thenReturn(Single.just<List<Want>>(emptyList()))
        whenever(controller.release).thenReturn(releaseWithLabelSomeForSale)

        presenter.fetchArtistDetails(id)
        testScheduler.triggerActions()

        verify(discogsInteractor).fetchReleaseDetails(id)
        verify(controller).setReleaseLoading(true)
        verify(discogsInteractor).fetchLabelDetails(id)
        verify(daoManager).storeViewedRelease(releaseWithLabelSomeForSale, artistsBeautifier)
        verify(controller).release = releaseWithLabelSomeForSale
        verify(controller).setReleaseListings(listArgumentCaptor.capture() as MutableList<ScrapeListing>?)
        assertEquals(listArgumentCaptor.value.size, 4)
        verify(discogsInteractor, times(1)).checkIfInCollection(controller, releaseWithLabelSomeForSale)
        verify(discogsInteractor, times(1)).checkIfInWantlist(controller, releaseWithLabelSomeForSale)
        verify(discogsInteractor).getReleaseMarketListings(id)
        verify(controller, times(2)).release
        verify(controller).setCollectionWantlistChecked(true)
        verify(controller).setListingsLoading(true)
    }
}